cpp/include/raft/util/reduction.cuh

```diff
@@ -0,0 +1,268 @@
+  return val;
+}
+
+/**
+ * @brief Warp-level reduction
+ * @tparam T Value type to be reduced
+ * @tparam ReduceLambda Reduction operation type
+ * @param val input value
+ * @param reduce_op Reduction operation
+ * @return Reduction result. All lanes will have the valid result.
+ * @note Why not cub? Because cub doesn't seem to allow working with arbitrary
+ *       number of warps in a block. All threads in the warp must enter this
+ *       function together
+ */
+template <typename T, typename ReduceLambda>
+DI T warpReduce(T val, ReduceLambda reduce_op)
+{
+  return logicalWarpReduce<WarpSize>(val, reduce_op);
+}
+
+/**
+ * @brief Warp-level reduction
+ * @tparam T Value type to be reduced
+ * @param val input value
+ * @return Reduction result. All lanes will have the valid result.
+ * @note Why not cub? Because cub doesn't seem to allow working with arbitrary
+ *       number of warps in a block. All threads in the warp must enter this
+ *       function together
+ */
+template <typename T>
+DI T warpReduce(T val)
+{
+  return warpReduce(val, raft::add_op{});
+}
+
+/**
+ * @brief 1-D block-level reduction
+ * @param val input value
+ * @param smem shared memory region needed for storing intermediate results. It
+ *             must alteast be of size: `sizeof(T) * nWarps`
+ * @param reduce_op a binary reduction operation.
+ * @return only the thread0 will contain valid reduced result
+ * @note Why not cub? Because cub doesn't seem to allow working with arbitrary
+ *       number of warps in a block. All threads in the block must enter this
+ *       function together. cub also uses too many registers
+ */
+template <typename T, typename ReduceLambda = raft::add_op>
+DI T blockReduce(T val, char* smem, ReduceLambda reduce_op = raft::add_op{})
+{
+  auto* sTemp = reinterpret_cast<T*>(smem);
+  int nWarps  = (blockDim.x + WarpSize - 1) / WarpSize;
+  int lid     = laneId();
+  int wid     = threadIdx.x / WarpSize;
+  val         = warpReduce(val, reduce_op);
+  if (lid == 0) sTemp[wid] = val;
+  __syncthreads();
+  val = lid < nWarps ? sTemp[lid] : T(0);
+  return warpReduce(val, reduce_op);
+}
+
+/**
+ * @brief warp-level weighted selection of an index.
+ * It selects an index with the given discrete probability
+ * distribution(represented by weights of each index)
+ * @param rng random number generator, must have next_u32() function
+ * @param weight weight of the rank/index.
+ * @param idx index to be used as rank
+ * @return only the thread0 will contain valid reduced result
+ */
+template <typename T, typename rng_t, typename i_t = int>
+DI T warpWeightedSelect(rng_t& rng, T& weight, i_t& idx)
+{
+  static_assert(std::is_integral<T>::value, "The type T must be an integral type.");
+#pragma unroll
+  for (i_t offset = raft::WarpSize / 2; offset > 0; offset /= 2) {
+    T tmp_weight = shfl(weight, laneId() + offset);
+    i_t tmp_idx  = shfl(idx, laneId() + offset);
+    T sum        = (tmp_weight + weight);
+    weight       = sum;
```

Comment:
It would be good to add some developer notes, something along the lines:
```
Let w_i be the weight stored on thread i.  We calculate the cumulative distribution function
F_i = \sum_{k=0..i} weight_i.

Sequentially, we could select one of the elements with with the desired probability using the following method.

We can consider that each element has a subinterval assigned: [F_{i-1}, F_i). We generate a uniform random number in the [0, F_i) range, and check which subinterval it falls. We return idx corresponding to the selected subinterval.

In parallel, we do a tree reduction and make a selection at every step when we combine two values.
```

In every step we need to shuffle both weight and idx. Is this actually faster than doing a separate inclusive scan on the weights followed by a selection?

Reply:
The comment is clear, thanks for writing it :) 

I am not sure we probably need to benchmark it but scan uses two times the number of steps as reduction. Additionally, do we have efficient inclusive scan functions on block and warp level ? I have one inside cuOpt but need to tidy up and create a separate PR for that.

Reply:
> we probably need to benchmark it but scan uses two times the number of steps as reduction

I think for this PR it is fine to keep the current form and add a todo item to return to this question later.